parsers/US_SPP.py

```diff
@@ -0,0 +1,198 @@
+#!usr/bin/env python3
+
+"""Parser for the Southwest Power Pool area of the United States."""
+
+from dateutil import parser, tz
+from io import StringIO
+from logging import getLogger
+import pandas as pd
+import requests
+
+generation_url = 'https://marketplace.spp.org/public-data-api/gen-mix/asFile'
+
+exchange_url = 'https://marketplace.spp.org/public-data-api/interchange-trend/asFile'
+
+mapping = {'coal': 'coal',
+           'unknown': 'unknown',
+           'Wind': 'wind',
+           'Nuclear': 'nuclear',
+           'Hydro': 'hydro',
+           'Solar': 'solar',
+           'Natural Gas': 'gas',
+           'Diesel Fuel Oil': 'oil',
+           'Waste Disposal Services': 'biomass'
+            }
+
+tie_mapping = {'US-MISO->US-SPP': ['AMRN', 'DPC', 'GRE', 'MDU', 'MEC', 'NSP', 'OTP']}
+
+# NOTE
+# Data sources return timestamps in GMT.
+# Energy storage situation unclear as of 16/03/2018, likely to change quickly in future.
+
+
+def get_data(url, session=None):
+    """Returns a pandas dataframe."""
+
+    s=session or requests.Session()
+    req = s.get(url, verify=False)
+    df = pd.read_csv(StringIO(req.text))
+
+    return df
+
+
+def data_processor(df, logger):
+    """
+    Takes a dataframe and logging instance as input.
+    Checks for new generation types and logs awarning if any are found.
+    Parses the dataframe row by row removing unneeded keys.
+    Returns a list of 2 element tuples, each containing a datetime object
+    and production dictionary.
+    """
+
+    # Remove leading whitespace in column headers.
+    df.columns = df.columns.str.strip()
+
+    keys_to_remove = ['Coal Market', 'Coal Self', 'GMT MKT Interval', 'Average Actual Load',
+                      'Other', 'Waste Heat']
+
+    # Check for new generation columns.
+    known_keys = list(mapping.keys()) + keys_to_remove
+    column_headers = list(df)
+
+    unknown_keys = []
+    for heading in column_headers:
+        if heading not in known_keys:
+            logger.warning('New column \'{}\' present in US-SPP data source.'.format(
+                heading), extra={'key': 'US-SPP'})
+            unknown_keys.append(heading)
+
+    keys_to_remove += unknown_keys
+
+    processed_data = []
+    for index, row in df.iterrows():
+        production = row.to_dict()
+        production['coal'] = production['Coal Market'] + production['Coal Self']
+
+        extra_unknowns = sum([production[k] for k in unknown_keys])
+        production['unknown'] = production['Other'] + production['Waste Heat'] + extra_unknowns
+
+        dt_aware = parser.parse(production['GMT MKT Interval'])
+
+        for k in keys_to_remove:
+            production.pop(k, None)
+
+        mapped_production = {mapping.get(k,k):v for k,v in production.items()}
```

Comment:
I think at this point we've assured that `production` only contains keys that are in `mapping`, so the default value in `mapping.get(k, k)` is not needed?

Reply:
With the 'coal' and 'unknown' keys gone from MAPPING, get stops them raising KeyErrors in the dict comprehension. So I think it needs to stay.